# Exercise 1: Advertisement agency sensitivity analysis

## Modeling


**Decision variables **
Let $i=\{1,2,3,4\}$ be in the index for the advertisers such that
* $i=1$ is GEICO
* $i=2$ is Delta
* $i=3$ is T-Mobile
* $i=4$ is Capital One

Let $j=\{1,2\}$ be the index for the category, such as $j=1$ is the Sports categorty and $j=2$ is the National category.

So $x_{11}$ is the number of impressions for Geico in Sports and $x_{12}$ is the number of impressions for GEICO int he National category


**Objective**

The objective of the platform, i.e., Washington Post, is to maximize advertising revenues. These revenues are driven by the cost per click charged to the advertisers, i.e., $\pi=\$2.5$ and the click-through rates (CTR).

Let $\kappa_{ij}$ be the CTR of advertiser $i$ in category $j$. For instance, the CTR of T-Mobile in the sports category is $\kappa_{31}=1.0\%$, whereas its CTR in the National category is $\kappa_{32}=3.0\%$.

The objective function is thus:
\begin{equation}
Maximize \quad 2.3\times \sum_{i=1}^4\sum_{j=1}^{2}  \kappa_{ij}\times x_{ij}
\end{equation}





**Constraints**
The first table provides the following constraints

\begin{align}
x_{11}&\geq 2,000,000\\
x_{12}&\geq 1,000,000\\
x_{22}&\geq 1,000,000\\
x_{21}+x_{22}&\geq 2,000,000\\
x_{31}&\geq 1,000,000\\
x_{32}&\geq 1,000,000\\
x_{31}+x_{32}&\geq 3,000,000\\
x_{41}+x_{42}&\geq 2,000,000\\
\end{align}

"Assume that the Sports section gets six million views per day and the National section
get five million views per day"
This statement gives two constraints:
\begin{align}
\sum_{i=1}^{4}x_{i1}&\leq 6,000,000\\
\sum_{i=1}^{4}x_{i2}&\leq 5,000,000\\
\end{align}

Non-negativity constraints are
\begin{equation}
x_{ij}\geq 0
\end{equation}

### Import Libraries

In [0]:
import math
import numpy
import cvxpy as cvx
import numpy as np

### Variable declaration

In [0]:
# number of impressions required in each section by each company
x11 = cvx.Variable(1, integer=True)# Geico Sports
x12 = cvx.Variable(1, integer=True)# Geico National
x21 = cvx.Variable(1, integer=True)# Delta Sports
x22 = cvx.Variable(1, integer=True)# Delta National
x31 = cvx.Variable(1, integer=True)# T-Mobile Sports
x32 = cvx.Variable(1, integer=True)# T-Mobile National
x41 = cvx.Variable(1, integer=True)# Capital One Sports
x42 = cvx.Variable(1, integer=True)# Capital One National

### Constraints

In [0]:
c1= x11>= 2000000 
c2= x12>= 1000000 
c3=x22>= 1000000 
c4= x21+x22== 2000000 
c5=x31>=1000000
c6=x32>=1000000
c7=x31+x32==3000000
c8=x41+x42==2000000
c9=x11+x21+x31+x41<=6000000
c10=x12+x22+x32+x42<=5000000
c11=x21>=0
c12=x41>=0
c13=x42>=0
con=[c1,c2,c3,c4,c5,c6,c7,c8,c9,c10,c11,c12,c13]

### Objective function

In [0]:
#objective
k11=2.5/100
k12=0.8/100
k21=2.0/100
k22=1.0/100
k31=1.0/100
k32=3.0/100
k41=1.5/100
k42=2.0/100
OF=(k11*x11 + k12*x12 + k21*x21 + k22*x22 + k31*x31 + k32*x32 + k41*x41 + k42*x42) * 2.3
objective = cvx.Maximize(OF)

### Maximum Revenue

In [10]:

prob = cvx.Problem(objective, con)
result = prob.solve()

print('optimal revenue')
print (int(np.round(prob.value)))

optimal revenue
501400


### Impressions allocation

In [6]:
from prettytable import PrettyTable
    
x = PrettyTable()

x.field_names = ["Advertisers", "Sports", "National"]

x.add_row(["Geico", int(np.round(x11.value)), int(np.round(x12.value))])
x.add_row(["Delta", int(np.round(x21.value)), int(np.round(x22.value))])
x.add_row(["T-Mobile", int(np.round(x31.value)), int(np.round(x32.value))])
x.add_row(["Capital One", int(np.round(x41.value)), int(np.round(x42.value))])

print(x)

+-------------+---------+----------+
| Advertisers |  Sports | National |
+-------------+---------+----------+
|    Geico    | 3000000 | 1000000  |
|    Delta    | 1000000 | 1000000  |
|   T-Mobile  | 1000000 | 2000000  |
| Capital One | 1000000 | 1000000  |
+-------------+---------+----------+


## Sensitivity Analysis

We are going to perform sensitivity analysis based on changing the Click Through Rate (CTR). 

**Changing one advetiser's Click Through Rate for both sections**

In [0]:
# We take the midpoint of the range of values for each advertiser's each section
k11=2.5/100
k12=0.6/100
k21=2.0/100
k22=1.0/100
k31=1.0/100
k32=3.0/100
k41=1.75/100
k42=2.0/100

# Before running sensitivity analysis, we are saving the original allocation results
# this allocation will be used to compare how sensitive the allocation is with respect to different CTR
opt_x11 = int(np.round(x11.value))
opt_x12 = int(np.round(x12.value))
opt_x21 = int(np.round(x21.value))
opt_x22 = int(np.round(x22.value))
opt_x31 = int(np.round(x31.value))
opt_x32 = int(np.round(x32.value))
opt_x41 = int(np.round(x41.value))
opt_x42 = int(np.round(x42.value))

# compares the allocation with the optimal allocation (as a benchmark) arrived in the above section.
# if each advertiser's each section matches our benchmark value, we return True else False
# returns True/False
def compare_difference(x11, x12, x21, x22, x31, x32, x41, x42):
  compare_result =  (int(np.round(x11.value)) != opt_x11) or (int(np.round(x12.value)) != opt_x12) or (int(np.round(x21.value)) != opt_x21) or (int(np.round(x22.value)) != opt_x22) or (int(np.round(x31.value)) != opt_x31) or (int(np.round(x32.value)) != opt_x32) and (int(np.round(x41.value)) != opt_x41) and (int(np.round(x42.value)) != opt_x42)  
  return compare_result

In [12]:
print("Sensitivity analysis for different CTR. Printing only the allocation that are different from the benchmark allocation")

# on an increment of 0.5, the different CTR values Geiko Sports can take
k11_values = [2.0/100, 2.5/100, 3.0/100]
# on an increment of 0.5, the different CTR values Geiko National can take
k12_values = [0.1/100, 0.6/100, 1.1/100]

optimal_geiko = []
optimal_delta = []
optimal_tmobile = []
optimal_capitalone = []

for k11 in k11_values:
  
  for k12 in k12_values:
    print("CTR Geiko Sports: %s" %k11)
    print("CTR Geiko National: %s" %round(k12, 2))
    OF=(k11*x11 + k12*x12+ k21*x21 + k22*x22 + k31*x31+ k32*x32+k41*x41+k42*x42)*2.3
    objective = cvx.Maximize(OF)

    #solving
    prob = cvx.Problem(objective, con)
    result = prob.solve()

    print('optimal revenue')
    print (int(np.round(prob.value)))
    optimal_geiko.append(int(np.round(prob.value)))
    
    x = PrettyTable()

    x.field_names = ["Advertisers", "Sports", "National"]

    x.add_row(["Geico", int(np.round(x11.value)), int(np.round(x12.value))])
    x.add_row(["Delta", int(np.round(x21.value)), int(np.round(x22.value))])
    x.add_row(["T-Mobile", int(np.round(x31.value)), int(np.round(x32.value))])
    x.add_row(["Capital One", int(np.round(x41.value)), int(np.round(x42.value))])
    
    # we are printing the allocation only when it is different from benchmark
    if compare_difference(x11, x12, x21, x22, x31, x32, x41, x42):
      print(x)

      
# after finishing the loop, replacing with the original values      
k11=2.5/100
k12=0.6/100

# on an increment of 0.5, the different CTR values Delta Sports can take
k21_values = [1.5/100, 2.0/100, 2.5/100]
# on an increment of 0.5, the different CTR values Delta National can take
k22_values = [0.5/100, 1.0/100, 1.5/100]

for k21 in k21_values:
  for k22 in k22_values:
    print("CTR Delta Sports: %s" %k21)
    print("CTR Delta National: %s" %round(k22, 2))
    OF=(k11*x11 + k12*x12+ k21*x21 + k22*x22 + k31*x31+ k32*x32+k41*x41+k42*x42)*2.3
    objective = cvx.Maximize(OF)

    #solving
    prob = cvx.Problem(objective, con)
    result = prob.solve()

    print('optimal revenue')
    print (int(np.round(prob.value)))
    optimal_delta.append(int(np.round(prob.value)))
    x = PrettyTable()

    x.field_names = ["Advertisers", "Sports", "National"]

    x.add_row(["Geico", int(np.round(x11.value)), int(np.round(x12.value))])
    x.add_row(["Delta", int(np.round(x21.value)), int(np.round(x22.value))])
    x.add_row(["T-Mobile", int(np.round(x31.value)), int(np.round(x32.value))])
    x.add_row(["Capital One", int(np.round(x41.value)), int(np.round(x42.value))])

    # we are printing the allocation only when it is different from benchmark
    if compare_difference(x11, x12, x21, x22, x31, x32, x41, x42):
      print(x)

# after finishing the loop, replacing with the original values      
k21=2.0/100
k22=1.0/100

# on an increment of 0.5, the different CTR values T-mobile Sports can take
k31_values = [0.5/100, 1.0/100, 1.5/100]
# on an increment of 0.5, the different CTR values T-mobile National can take
k32_values = [2.5/100, 3.0/100, 3.5/100]

for k31 in k31_values:
  
  for k32 in k32_values:
    print("CTR T-mobile Sports: %s" %round(k31, 2))
    print("CTR T-mobile National: %s" %round(k32, 2))
    OF=(k11*x11 + k12*x12+ k21*x21 + k22*x22 + k31*x31+ k32*x32+k41*x41+k42*x42)*2.3
    objective = cvx.Maximize(OF)

    #solving
    prob = cvx.Problem(objective, con)
    result = prob.solve()

    print('optimal revenue')
    print (int(np.round(prob.value)))
    optimal_tmobile.append(int(np.round(prob.value)))
    x = PrettyTable()

    x.field_names = ["Advertisers", "Sports", "National"]

    x.add_row(["Geico", int(np.round(x11.value)), int(np.round(x12.value))])
    x.add_row(["Delta", int(np.round(x21.value)), int(np.round(x22.value))])
    x.add_row(["T-Mobile", int(np.round(x31.value)), int(np.round(x32.value))])
    x.add_row(["Capital One", int(np.round(x41.value)), int(np.round(x42.value))])

    # we are printing the allocation only when it is different from benchmark
    if compare_difference(x11, x12, x21, x22, x31, x32, x41, x42):
      print(x)

# after finishing the loop, replacing with the original values      
k31=1.0/100
k32=3.0/100

# on an increment of 0.5, the different CTR values Capital One Sports can take
k41_values = [1.5/100, 2.0/100]
# on an increment of 0.5, the different CTR values Capital One National can take
k42_values = [1.5/100, 2.0/100, 2.5/100]

for k41 in k41_values:
  
  for k42 in k42_values:
    print("CTR Captial One Sports: %s" %k41)
    print("CTR Captial One National: %s" %k42)
    OF=(k11*x11 + k12*x12+ k21*x21 + k22*x22 + k31*x31+ k32*x32+k41*x41+k42*x42)*2.3
    objective = cvx.Maximize(OF)

    #solving
    prob = cvx.Problem(objective, con)
    result = prob.solve()
    
    print('optimal revenue')
    print (int(np.round(prob.value)))
    optimal_capitalone.append(int(np.round(prob.value)))
    x = PrettyTable()

    x.field_names = ["Advertisers", "Sports", "National"]

    x.add_row(["Geico", int(np.round(x11.value)), int(np.round(x12.value))])
    x.add_row(["Delta", int(np.round(x21.value)), int(np.round(x22.value))])
    x.add_row(["T-Mobile", int(np.round(x31.value)), int(np.round(x32.value))])
    x.add_row(["Capital One", int(np.round(x41.value)), int(np.round(x42.value))])
    
    # we are printing the allocation only when it is different from benchmark
    if compare_difference(x11, x12, x21, x22, x31, x32, x41, x42):
      print(x)

# after finishing the loop, replacing with the original values      
k41=1.5/100
k42=2.0/100


Sensitivity analysis for different CTR. Printing only the allocation that are different from the benchmark allocation
CTR Geiko Sports: 0.02
CTR Geiko National: 0.0
optimal revenue
456550
CTR Geiko Sports: 0.02
CTR Geiko National: 0.01
optimal revenue
468050
CTR Geiko Sports: 0.02
CTR Geiko National: 0.01
optimal revenue
479550
CTR Geiko Sports: 0.025
CTR Geiko National: 0.0
optimal revenue
491050
CTR Geiko Sports: 0.025
CTR Geiko National: 0.01
optimal revenue
502550
CTR Geiko Sports: 0.025
CTR Geiko National: 0.01
optimal revenue
514050
CTR Geiko Sports: 0.03
CTR Geiko National: 0.0
optimal revenue
525550
CTR Geiko Sports: 0.03
CTR Geiko National: 0.01
optimal revenue
537050
CTR Geiko Sports: 0.03
CTR Geiko National: 0.01
optimal revenue
548550
CTR Delta Sports: 0.015
CTR Delta National: 0.01
optimal revenue
479550
CTR Delta Sports: 0.015
CTR Delta National: 0.01
optimal revenue
491050
CTR Delta Sports: 0.015
CTR Delta National: 0.01
optimal revenue
502550
CTR Delta Sports: 0.02
CTR 

### Section I: We are observing if the allocation changes, when the CTR changes.

![alt text](https://raw.github.com/krithikaceg/Optimization/master/Picture2.png)


As we can see form the above graph, Geiko is more sensitive to CTR rate changes.Whereas, t-mobile and Capital One is less sensitive.

### Section II: We are observing if the allocation changes, when the CTR changes.

At a time, we are changing the Click Through Rate of both Sports and National section.

As we can observe from the above result, the allocation changes only for the following CTR combination.

So, we can conclude that our allocation is not very sensitive to CTR.

In [50]:
x = PrettyTable()

x.field_names = ["Advertisers", "Sports", "National"]

x.add_row(["Geico", 3000000, 1000000])
x.add_row(["Delta", 928571, 1071429])
x.add_row(["T-Mobile", 1000000, 2000000])
x.add_row(["Capital One", 1071429, 928571])
print(x)
    
    
x = PrettyTable()

x.field_names = ["Advertisers", "Sports", "National"]

x.add_row(["Geico", 0.025, 0.06])
x.add_row(["Delta", 0.02, 0.01])
x.add_row(["T-Mobile", 0.01, 0.03])
x.add_row(["Capital One", 0.0175, 0.02])
print(x)

+-------------+---------+----------+
| Advertisers |  Sports | National |
+-------------+---------+----------+
|    Geico    | 3000000 | 1000000  |
|    Delta    |  928571 | 1071429  |
|   T-Mobile  | 1000000 | 2000000  |
| Capital One | 1071429 |  928571  |
+-------------+---------+----------+
+-------------+--------+----------+
| Advertisers | Sports | National |
+-------------+--------+----------+
|    Geico    | 0.025  |   0.06   |
|    Delta    |  0.02  |   0.01   |
|   T-Mobile  |  0.01  |   0.03   |
| Capital One | 0.0175 |   0.02   |
+-------------+--------+----------+


To The Washington Post,

Based on our analysis, we have suggested optimal allocation of impressions for each company in each division. Based on the past history of advertisers and their CTRs, we have also come up with the maximum profit the company would achieve. 

However, this optimal allocation is calculated based on deterministic model. Since things vary over time, we have accounted for the changes in CTR. We have performed sensitivity analysis on the chnage of CTRfor every advertiser for each section.

Our Type I sensitivity analysis suggests that, Geiko is more sensitive to CTR changes. WHen CTR rates go from 2% to 3% in Sports and 0.1% to 1.1% in National, the optimal revenue obtained is varying a lot. We suggest you made need to pay careful attention when CTR rates of Geiko changes. 

On the other hand, our Type II sensitivity analysis suggests that even though the CTR rates change, the allocations for each advertiser's each section doesn't vary much. To get into details you can refer to "Section II: We are observing if the allocation changes, when the CTR changes." 

Based on the historical data of The Washington Post and the constraints you have given, we have created optimization model and performed optimization. However, this is only a suggestion from our side to allocate the impressions. It is finally the decision of your company to go ahead with it or not. As things change over time, we can only expect to happen in the way we predicted, nonetheless, we cannot assure it will be exactly the same.

Please come back to us in case of any queries or further requirement.


# Question 2: Operations Excellence of plants

You operate two plants, i.e., A and B. Each plant makes two products, “standard” and “deluxe”. A unit of
standard gives a profit contribution of \\$10, while a unit of deluxe gives a profit contribution of $15.

Each plant uses two processes, grinding and polishing, for producing its products. Plant A has a grinding capacity of 80 hours per week and polishing capacity of 60 hours per week. For plant B, these capacities are 60 and 75 hours per week, respectively.
The grinding and polishing times in hours for a unit of each type of product in each factory are as given below.

For Plant A
Standard Deluxe
Grinding 4 2
Polishing 2 5
For Plant B
Standard Deluxe
Grinding 5 3
Polishing 5 6

It is possible, for example, that plant B has older machines than plant A, resulting in higher unit processing times. In addition, each unit of each product uses 4 kg of a raw material, which we refer to as raw. The company has 120 kg of raw available per week. To start with, we will assume that plant A is allocated 75 kg of raw per week and plant B the remaining 45 kg per week. Each plant can build a very simple linear programming model to maximize its profit contribution.

### Import Libraries

In [0]:
import math
import numpy
import cvxpy as cvx
import numpy
from prettytable import PrettyTable

### Variable declaration

In [0]:
# number of products produced by each plant
a1 = cvx.Variable(1, integer=True)# Standard from Plant A
a2 = cvx.Variable(1, integer=True)# Deluxe from Plant A
b1 = cvx.Variable(1, integer=True)# Standard from Plant B
b2 = cvx.Variable(1, integer=True)# Deluxe from Plant B

### Constraints

In [0]:
c1= a1 >= 0
c2= a2 >= 0 
c3= b1 >= 0 
c4= b2 >= 0 
c5= a1*4 + a2*2 <= 80
c6= a1*2 + a2*5 <= 60
c7= b1*5 + b2*3 <= 60
c8= b1*5 + b2*6 <= 75
c9= a1*4 + a2*4 <= 75
c10= b1*4 + b2*4 <= 45

con_A = [c1,c2,c5,c6,c9]
con_B = [c3,c4,c7,c8,c10]
con = [c1,c2,c3,c4,c5,c6,c7,c8,c9,c10]

### Objective

In [0]:
#objective

std_price = 10 # price of standard product
dlx_price = 15 # price of deluxe product

OF_A = a1*std_price + a2*dlx_price
OF_B = b1*std_price + b2*dlx_price
OF = a1*std_price + a2*dlx_price + b1*std_price + b2*dlx_price

objective_A = cvx.Maximize(OF_A)
objective_B = cvx.Maximize(OF_B)
objective = cvx.Maximize(OF)



### Plant A: Maximum Revenue & Allocation

In [13]:
#solving - Plant A
prob_A = cvx.Problem(objective_A, con_A)
result_A = prob_A.solve()
print('\nPLANT A')

print('optimal revenue')
print (int(np.round(prob_A.value)))
    
xa = PrettyTable()

xa.field_names = ["Product", "Plant A"]

xa.add_row(["Standard", int(np.round(a1.value))])
xa.add_row(["Deluxe", int(np.round(a2.value))])

print(xa)




PLANT A
optimal revenue
220
+----------+---------+
| Product  | Plant A |
+----------+---------+
| Standard |    10   |
|  Deluxe  |    8    |
+----------+---------+


To attain a maximum profit in Plant A, it should produce ten Standard products and eight Deluxe products. This would yield in a revenue of $220.

### Plant B: Maximum Revenue & Allocation

In [12]:
#solving - Plant B
prob_B = cvx.Problem(objective_B, con_B)
result_B = prob_B.solve()

print('\nPLANT B')
print('optimal revenue')
print (int(np.round(prob_B.value)))

xb = PrettyTable()

xb.field_names = ["Product", "Plant B"]

xb.add_row(["Standard", int(np.round(b1.value))])
xb.add_row(["Deluxe", int(np.round(b2.value))])
print(xb)


PLANT B
optimal revenue
165
+----------+---------+
| Product  | Plant B |
+----------+---------+
| Standard |    0    |
|  Deluxe  |    11   |
+----------+---------+


To attain a maximum profit in Plant B, it should produce eleven Deluxe products. This would yield in a revenue of $165.

### Analysis I: Plant A and Plant B: Maximum Revenue & Allocation

In [14]:
#solving - Plant A and B
prob = cvx.Problem(objective, con)
result = prob.solve()
print('\nPLANT A & B')
print('optimal revenue')
print (int(np.round(prob.value)))

x = PrettyTable()

x.field_names = ["Product", "Plant A", "Plant B"]

x.add_row(["Standard", int(np.round(a1.value)), int(np.round(b1.value))])
x.add_row(["Deluxe", int(np.round(a2.value)), int(np.round(b2.value))])
print(x)


PLANT A & B
optimal revenue
385
+----------+---------+---------+
| Product  | Plant A | Plant B |
+----------+---------+---------+
| Standard |    10   |    0    |
|  Deluxe  |    8    |    11   |
+----------+---------+---------+


Compared to the solution from individual plant profit contribution, the allocation are more or less same if they do not share raw material between them.


### Analysis II: Alternative Solution - sharing raw material

In [15]:
#solving - Plant A and B

# new constraint - sharing the raw material between A and B instead of allocating fixed amount to each plant
c11 = 4*(a1+a2+b1+b2) <= 120
con = [c1,c2,c3,c4,c5,c6,c7,c8,c11]
prob = cvx.Problem(objective, con)
result = prob.solve()
print('\nPLANT A & B')
print('optimal revenue')
print (int(np.round(prob.value)))

x = PrettyTable()

x.field_names = ["Product", "Plant A", "Plant B"]

x.add_row(["Standard", int(np.round(a1.value)), int(np.round(b1.value))])
x.add_row(["Deluxe", int(np.round(a2.value)), int(np.round(b2.value))])
print(x)



PLANT A & B
optimal revenue
400
+----------+---------+---------+
| Product  | Plant A | Plant B |
+----------+---------+---------+
| Standard |    10   |    0    |
|  Deluxe  |    8    |    12   |
+----------+---------+---------+


If they share raw material, their revenue increases from \\$385 to $400.



### Analysis III: Better Solution - Unlimited supply of raw material

In [16]:

#solving - Plant A and B
# removing the constraint on raw material
# if company is able to provide infinite amount of raw material to both plants
con = [c1,c2,c3,c4,c5,c6,c7,c8]
prob = cvx.Problem(objective, con)
result = prob.solve()
print('\nPLANT A & B')
print('optimal revenue')
print (int(np.round(prob.value)))

x = PrettyTable()

x.field_names = ["Product", "Plant A", "Plant B"]

x.add_row(["Standard", int(np.round(a1.value)), int(np.round(b1.value))])
x.add_row(["Deluxe", int(np.round(a2.value)), int(np.round(b2.value))])
print(x)


PLANT A & B
optimal revenue
425
+----------+---------+---------+
| Product  | Plant A | Plant B |
+----------+---------+---------+
| Standard |    17   |    0    |
|  Deluxe  |    5    |    12   |
+----------+---------+---------+


Based on the constraints we have received, we have created our optimization model and perofrmed optimization. We have done three types of analysis pertaining to different conditions and I would love to go over them one by one. Each of the analysis differes the way raw material is distributed among the two plants A and B.

In our first analysis, we have created our constraints based on "strict allocation." When Plant A and B are given 75 and 45 raw material each, we are able to see a profit of $385. Please refer to Analysis I for the allocation details.

In our second analysis, we have created our constraints based on "sharing resources." When Plant A and B are together given 120 amount of raw material we are able to see a profit of $400. Please refer to Analysis II for the allocation details. 

In our final analysis, we have created our constraints based on "unlimited supply." When Plant A and B are given unlimited supply of raw materials, we are able to see a profit of $425. Please refer to Analysis III for the allocation details.

We ran the above three analysis to find the optimal solution using our optimization model by changing our constraints. If you can purchase 136 quantities of raw materials and give 68 to each plant A and Plant B, they can make complete use of the resources allocated to them and be return maximum profit. 

Let us know how much more raw materials you could purchase, so that we can run our model with the new constraints and provide better results.

# Question 3: Hotel La Quinta Motor Inns (LQM)

La Quinta Motors are planning to purchase some hotels in California. They have a budget of 10 million. They want to purchase hotels that would yield them maximum profit. 

Hotels have different prices and profitabilities. We need to suggest an optimal solution LQM that suggests which hotels they should buy maximising their profit.

### Import libraries

In [0]:
import pandas as pd
import numpy as np
import math
import cvxpy as cvx
from numpy import matrix 
from cvxpy import *
from prettytable import PrettyTable

### Reading input from file

Using the below formula to calculate profitability.

𝑃𝑟𝑜𝑓𝑖𝑡𝑎𝑏𝑖𝑙𝑖𝑡𝑦 = 39.05 − 5.41×𝑆𝑡𝑎𝑡𝑒 𝑃𝑜𝑝𝑢𝑙𝑎𝑡𝑖𝑜𝑛 𝑝𝑒𝑟 𝐼𝑛𝑛 + 
5.86×𝑃𝑟𝑖𝑐𝑒 𝑜𝑓 𝑡ℎ𝑒 𝐼𝑛𝑛
− 3.09×𝑆𝑞𝑢𝑎𝑟𝑒 𝑅𝑜𝑜𝑡 𝑜𝑓 𝑡ℎ𝑒 𝑀𝑒𝑑𝑖𝑎𝑛 𝐼𝑛𝑐𝑜𝑚𝑒 𝑖𝑛 𝑡ℎ𝑒 𝑎𝑟𝑒𝑎 + 
1.75×𝐶𝑜𝑙𝑙𝑒𝑔𝑒 𝑆𝑡𝑢𝑑𝑒𝑛𝑡𝑠 𝑖𝑛 𝑡ℎ𝑒 𝐴𝑟𝑒𝑎
                                    

Profitability is calculated in excel sheet that is attached. The rows are sorting with the profitability value in their descending order.


In [0]:
url = 'https://raw.github.com/krithikaceg/Optimization/master/SelectingHotels%20%20MSBA.xlsx'
input_df = pd.read_excel(url)
hotel_list = input_df['Hotel'].tolist()
price_list = input_df['Price'].tolist()
profitability_list = input_df['Profitability'].tolist()
city_list =  input_df['Location'].tolist()

# creating cvx variable
x = cvx.Variable(16, boolean=True) # define a vector of n Boolean variables, i.e., x_i={0,1}  

### Data computed from input file

In [4]:
input_df

Hotel                      Location     Price  Price (normalized)  \
0       2            Fresno, California  10000000            1.699076   
1       6        Long Beach, California   8950000            1.402123   
2       1            Eureka, California   2925000           -0.301823   
3       3            Fresno, California   3750000           -0.068503   
4       4            Fresno, California   3500000           -0.139206   
5      14  South Lake Tahoe, California   3750000           -0.068503   
6      12  South Lake Tahoe, California   2500000           -0.422019   
7      13  South Lake Tahoe, California   1975000           -0.570495   
8      10  South Lake Tahoe, California   1650000           -0.662409   
9      15  South Lake Tahoe, California   1475000           -0.711902   
10     11  South Lake Tahoe, California   1125000           -0.810886   
11     16  South Lake Tahoe, California    750000           -0.916941   
12      5            Fresno, California    325000           -1.037136   
13      9       Los Angeles, California   4900000            0.256732   
14      7       Los Angeles, California   1950000           -0.577566   
15      8       Los Angeles, California   1750000           -0.634128   

    Square Root of Median Income (normalized)  \
0                                   -0.408199   
1                                    0.657845   
2                                   -0.812780   
3                                   -0.408199   
4                                   -0.408199   
5                                   -0.791006   
6                                   -0.791006   
7                                   -0.791006   
8                                   -0.791006   
9                                   -0.791006   
10                                  -0.791006   
11                                  -0.791006   
12                                  -0.408199   
13                                   0.167686   
14                                   0.167686   
15                                   0.167686   

    College Students in Area (normalized)  \
0                                0.311669   
1                                0.483711   
2                               -0.536413   
3                                0.311669   
4                                0.311669   
5                               -0.594905   
6                               -0.594905   
7                               -0.594905   
8                               -0.594905   
9                               -0.594905   
10                              -0.594905   
11                              -0.594905   
12                               0.311669   
13                               3.106215   
14                               3.106215   
15                               3.106215   

    State Population Per Inn (normalized)  Profitability  
0                               -0.474279          53.38  
1                               -0.557278          49.10  
2                               -0.995987          44.24  
3                               -0.474279          43.02  
4                               -0.474279          42.61  
5                               -0.426851          42.36  
6                               -0.426851          40.29  
7                               -0.426851          39.42  
8                               -0.426851          38.88  
9                               -0.426851          38.59  
10                              -0.426851          38.01  
11                              -0.426851          37.39  
12                              -0.474279          37.34  
13                               3.106530          28.67  
14                               3.106530          23.78  
15                               3.106530          23.45

### Variables affecting profitability

Part 1.

From the above equation, 𝑃𝑟𝑖𝑐𝑒 𝑜𝑓 𝑡ℎ𝑒 𝐼𝑛𝑛 affects profitability positively, whereas 𝑆𝑡𝑎𝑡𝑒 𝑃𝑜𝑝𝑢𝑙𝑎𝑡𝑖𝑜𝑛 𝑝𝑒𝑟 𝐼𝑛𝑛 affect profotability negatively. While they make the biggest impact, other variables such as 𝐶𝑜𝑙𝑙𝑒𝑔𝑒 𝑆𝑡𝑢𝑑𝑒𝑛𝑡𝑠 𝑖𝑛 𝑡ℎ𝑒 𝐴𝑟𝑒𝑎 affects profotability positively, while 𝑆𝑞𝑢𝑎𝑟𝑒 𝑅𝑜𝑜𝑡 𝑜𝑓 𝑡ℎ𝑒 𝑀𝑒𝑑𝑖𝑎𝑛 𝐼𝑛𝑐𝑜𝑚𝑒 𝑖𝑛 𝑡ℎ𝑒 𝑎𝑟𝑒𝑎 affect the profitability negatively.

𝑃𝑟𝑖𝑐𝑒 𝑜𝑓 𝑡ℎ𝑒 𝐼𝑛𝑛 can directly contribute to the profit as long it is not over-priced. 

State population per inn = State population/ number of inns in that place.

Since all of these hotels are in California, state population is a constant value across all the hotels in California state. So, it is the number of inns in that place that distinguishes one place from the other. (Hotels belonging to the same place has same State population per inn value.) Hence, if the state population per inn value is more, the number of inns is less in that place.

The Number of inns in that place increases with increase in people staying in that place. When the number of inns in the state is less, that implies the number of people staying is less, hence the profitability is less and vice versa.



Part 2.
Based on the formula used in the excel, the highest profitability is from Hotel #2, Fresno, California, whereas the lowest profitability is from Hotel #8, Los Angeles, California.

In [34]:
print("Maximum Profitable hotel: %s" %hotel_list[np.argmax(profitability_list)])
#print(hotel_list[np.argmax(profitability_list)])
print("Minimum Profitable hotel: %s" %hotel_list[np.argmin(profitability_list)])

Maximum Profitable hotel: 2
Minimum Profitable hotel: 8


Part 3:
Using the greedy approach, we choose from the most profitable hotel until we run out of budget. 

In that case, if we are given 10million, we would choose only one hotel: Hotel #2.

Sum of predicted profitability is 53.38

If we are trying to maximize our total predicted profitability, is this "not" a good approach. If we were trying to maximise the average profitability of the hotels purchsed, we would want to minimise the variance between the profitability of the hotels.

If we were given 20 million, the hotels purchased yielding higher total predicted profitability would be, hotels numbers: 0, 1, 11

In [37]:
budget = 20000000
hotels = []
for i in range(len(price_list)):
  balance_budget = budget - price_list[i]
  if (balance_budget > 0):
    hotels.append(i)
    budget = balance_budget
print(hotels)

[0, 1, 11]


### Constraints

In [0]:
c1 = sum(price_list*x) <= 10000000
con = [c1]

### Objective Function

In [0]:
obj1 = cvx.Maximize(sum(profitability_list*x))

### Maximum Profitability

In [22]:
prob = cvx.Problem(obj1, con)
result = prob.solve()

print('Maximum profitability')
print(round(prob.value))

Maximum profitability
270.0


### Hotels Chosen

In [38]:
np.around(x.value)

array([-0., -0.,  0.,  0.,  0.,  0.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,
        0.,  0.,  0.])

In [24]:
pt = PrettyTable()

print('Hotels that are chosen')

pt.field_names = ["Hotel", "City"]
for i in range(16):
  if (x[i].value > 0.99):
    pt.add_row([hotel_list[i], city_list[i]])

print(pt)

Hotels that are chosen
+-------+------------------------------+
| Hotel |             City             |
+-------+------------------------------+
|   12  | South Lake Tahoe, California |
|   13  | South Lake Tahoe, California |
|   10  | South Lake Tahoe, California |
|   15  | South Lake Tahoe, California |
|   11  | South Lake Tahoe, California |
|   16  | South Lake Tahoe, California |
|   5   |      Fresno, California      |
+-------+------------------------------+


Yes, it definitely is better than greedy solution. While we are trying to stick within our budget, we are also trying to find the optimal solution that maximises our profit. 
With the greedy solution, we achieved a profit of 53.38, whereas with the optimization we could achieve 270 with the same budget.

Optimization helps us to invest our money in the best possible way that yields us higher profit.

### Restricting the number of hotels in each city

Since LQM has decided to buy utmost two hotels in any place, we are adding extra constrints to perform that check while arriving at the optimal solution. This extra checks would yield us lesser profit compared to the optimal solution without these constraints and it is totally sensible.

In [0]:
c2=((city_list[5]=="South Lake Tahoe, California")*x[5] + 
    (city_list[6]=="South Lake Tahoe, California")*x[6] + 
    (city_list[7]=="South Lake Tahoe, California")*x[7] + 
    (city_list[8]=="South Lake Tahoe, California")*x[8] + 
    (city_list[9]=="South Lake Tahoe, California")*x[9] + 
    (city_list[10]=="South Lake Tahoe, California")*x[10] + 
    (city_list[11]=="South Lake Tahoe, California")*x[11] 
   )<=2
c3=((city_list[0]=='Fresno, California')*x[0] + 
    (city_list[3]=='Fresno, California')*x[3] + 
    (city_list[4]=='Fresno, California')*x[4] + 
    (city_list[12]=='Fresno, California')*x[12] 
   )<=2
c4=((city_list[13]=='Los Angeles, California')*x[13] + 
    (city_list[14]=='Los Angeles, California')*x[14] + 
    (city_list[15]=='Los Angeles, California')*x[15] 
   )<=2
con = [c1, c2, c3, c4]

In [26]:
prob = cvx.Problem(obj1, con)
result = prob.solve()

pt = PrettyTable()

print('Hotels that are chosen')

pt.field_names = ["Hotel", "City"]
for i in range(16):
  if (x[i].value > 0.99):
    pt.add_row([hotel_list[i], city_list[i]])

print(pt)

print('Maximum profitability')
print(round(prob.value))

Hotels that are chosen
+-------+------------------------------+
| Hotel |             City             |
+-------+------------------------------+
|   1   |      Eureka, California      |
|   10  | South Lake Tahoe, California |
|   11  | South Lake Tahoe, California |
|   5   |      Fresno, California      |
|   7   |   Los Angeles, California    |
|   8   |   Los Angeles, California    |
+-------+------------------------------+
Maximum profitability
206.0


As expected, the profit from optimal solution with constraints is lesser than the one from optimal solution without constraints. Since it is a business decision that LQM has decided not to buy more that two hotels in a place, we suggest what those hotels should be.

To La Quinta Motors,

Greetings! Congratulations on your recent budget allocation to purchase new hotels.

Based on the data and the constraints you have given us, we have calculated the profitability. Please find the data computed from the excel sheet in the above section 'Data computed from input file.' This data is sorted based on the proftiability of the hotels.

As you can see, Hotel #2 is the most profitable one, but also the most expensive one. If you are planning to invest all your 10 million budget in purchasing that hotel, you end up in $53.38 profit. However, we have built our optimization model that will help you choose more number of hotels within 10 million budget resulting in maximum profit.Based on your different constraints, we have developed different models.

First, this model returned us the following hotels: 5, 10, 11, 12, 13, 15, 16. This combination will fit within 10 million and at the same time, they give the maximum expected total profit. Even though this is a good combination, we found that  majority of the hotels are being purchased in South Lake Tahoe, California.

After considering your suggestion, we have added an extra constraint that would restrict the number of hotels in any place to two. With this constarint, we have arrived at the following hotels: 1, 5, 7, 8, 10, 11. These hotels are present totally in four areas and we strongly believe that this will help LQM expand business in different places.

This is the best possible suggestion we have arrived using the constraints you have given and your budget. However, if your constarints or budget change, please keep us informed. We can help you anytime!





